File: DemoServer/AdditionalLanguages/python/staticIndexes/additionalSourcesIndex/additionalSourcesIndex.py
from __future__ import annotations
from typing import List
from demo_example import Example, RunParamsBase
from models import Product

#region Usings
from ravendb import AbstractIndexCreationTask
#endregion

class RunParams(RunParamsBase):
    def __init__(self, price: int = None):
        self.price = price

#region Demo
#region Step_1
class Products_ByPrice(AbstractIndexCreationTask):
#endregion

    #region Step_2
    class IndexEntry:
        def __init__(
            self,
            product_name: str = None,
            original_price: float = None,
            sale_price: float = None,
            profit_price: float = None,
        ):
            self.product_name = product_name
            self.original_price = original_price
            self.sale_price = sale_price
            self.profit_price = profit_price
    #endregion

    def __init__(self):
        #region Step_3
        super().__init__()
        self.map = (
            "docs.Products.Select(product => new {"
            "    product_name = product.Name,"
            "    original_price = product.PricePerUnit,"
            "    sale_price = DiscountUtils.CalcSalePrice(product.PricePerUnit),"
            "    profit_price = DiscountUtils.CalcProfitPrice(product.PricePerUnit)"
            "})"
        )
        #endregion
        
        #region Step_4
        self.additional_sources = {"DiscountLogic": ADDITIONAL_SOURCE}
        #endregion
#endregion

class AdditionalSourcesIndex(Example):
    def run(self, run_params: RunParams) -> List[DataToShow]:
        price = run_params.price or 5
        Products_ByPrice().execute(self.document_store_holder.store())

        #region Demo
        #region Step_5
        with self.document_store_holder.store().open_session() as session:
            low_cost_products = list(
                session.query_index_type(Products_ByPrice, Products_ByPrice.IndexEntry)
                .where_less_than("sale_price", price)
                .order_by("sale_price")
            )
        #endregion
        #endregion

        # Manipulate results to show because index fields are not stored..
        products_data = [
            DataToShow(
                item.Name,
                item.PricePerUnit,
                item.PricePerUnit - (item.PricePerUnit / 100 * 25),
                item.PricePerUnit + (item.PricePerUnit / 100 * 25),
                )
            for item in low_cost_products
        ]
        return products_data

#region Demo
#region Step_6
ADDITIONAL_SOURCE = (
    "public static class DiscountUtils"
    "{"
    "    public static decimal CalcSalePrice(decimal price)"
    "    {"
    "        return price - price / 100M * 25M;"
    "    }"
    "     "
    "    public static decimal CalcProfitPrice(decimal price)"
    "    {"
    "        return price + price / 100M * 25M;"
    "    }"
    "}"
)
#endregion
#endregion

class DataToShow:
    def __init__(
        self,
        product_name: str = None,
        original_price: float = None,
        sale_price: float = None,
        profit_price: float = None,
    ):
        self.product_name = product_name
        self.original_price = original_price
        self.sale_price = sale_price
        self.profit_price = profit_price

File: DemoServer/AdditionalLanguages/php/staticIndexes/additionalSourcesIndex/AdditionalSourcesIndex.php
<?php

namespace RavenDB\Demo\staticIndexes\additionalSourcesIndex;

//region Usings
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Indexes\AdditionalSourcesArray;
//endregion

use RavenDB\Demo\common\DocumentStoreHolder;
use RavenDB\Demo\common\models\Product;

//region Demo
//region Step_1
class Products_ByPrice extends AbstractIndexCreationTask
//endregion
{
    public function __construct()
    {
        parent::__construct();

        //region Step_2
        $this->map = "docs.Products.Select(product => new {" .
            "    ProductName = product.Name," .
            "    OriginalPrice = product.PricePerUnit," .
            "    SalePrice = DiscountUtils.CalcSalePrice(product.PricePerUnit)," .
            "    ProfitPrice = DiscountUtils.CalcProfitPrice(product.PricePerUnit)" .
            "})";
        //endregion

        //region Step_3
        $this->setAdditionalSources(["DiscountLogic" => self::ADDITIONAL_SOURCE]);
        //endregion
    }

    //region Step_4
    public const ADDITIONAL_SOURCE = "public static class DiscountUtils" .
        "{" .
        "    public static decimal CalcSalePrice(decimal price)" .
        "    {" .
        "        return price - price / 100M * 25M;" .
        "    }" .
        "     " .
        "    public static decimal CalcProfitPrice(decimal price)" .
        "    {" .
        "        return price + price / 100M * 25M;" .
        "    }" .
        "}";
    //endregion
}
//endregion

//region Demo
//region Step_5
class IndexEntry
{
    private ?string $productName = null;
    private ?float $originalPrice = null;
    private ?float $salePrice = null;
    private ?float $profitPrice = null;

    public function getOriginalPrice(): ?float
    {
        return $this->originalPrice;
    }

    public function setOriginalPrice(?float $originalPrice): void
    {
        $this->originalPrice = $originalPrice;
    }

    public function getSalePrice(): ?float
    {
        return $this->salePrice;
    }

    public function setSalePrice(?float $salePrice): void
    {
        $this->salePrice = $salePrice;
    }

    public function getProfitPrice(): ?float
    {
        return $this->profitPrice;
    }

    public function setProfitPrice(?float $profitPrice): void
    {
        $this->profitPrice = $profitPrice;
    }
}
//endregion
//endregion

class AdditionalSourcesIndex
{
    public function __invoke(RunParams $runParams): array
    {
        $price = $runParams->getPrice() ?? 5;

        //region Demo
        $lowCostProducts = [];
        
        //region Step_6
        $session = DocumentStoreHolder::getStore()->openSession();
        
        try {
            $lowCostProducts = $session->query(Product::class, Products_ByPrice::class)
                ->whereLessThan("SalePrice", $price)
                ->orderBy("SalePrice")
                ->toList();
                
        } finally {
            $session->close();
        }
        //endregion
        //endregion

        // Manipulate results to show because index fields are Not stored.
        $productsData = [];
        
        /** @var Product $item */
        foreach ($lowCostProducts as $item) {
            $dataToShow = new DataToShow();
            $dataToShow->setProductName($item->getName());
            $dataToShow->setOriginalPrice($item->getPricePerUnit());
            $dataToShow->setSalesPrice($item->getPricePerUnit() - $item->getPricePerUnit() / 100 * 25);
            $dataToShow->setProfitPrice($item->getPricePerUnit() + $item->getPricePerUnit() / 100 * 25);
            $productsData[] = $dataToShow;
        }

        return $productsData;
    }
}
